Spatial.gd

```diff
@@ -490,101 +498,89 @@ func init_category_tree():
-		_waypoint_categories_to_godot_nodes(null, category, category.get_name(), false)
+	for waypoint_category in self.waypoint_data.get_category():
+		_waypoint_categories_to_godot_nodes(null, waypoint_category, self.categories, self.minimap, false)
 
 
-func _waypoint_categories_to_godot_nodes(item: TreeItem, category, full_category_name: String, collapsed: bool):
-	var category_item: TreeItem = self.marker_packs.create_item(item)
-	if category.get_name() == "":
-		# If this is called, there is an error in the Waypoint data
-		category_item.set_text(0, "No name")
-		category_item.set_metadata(0, "")
+func _waypoint_categories_to_godot_nodes(item: TreeItem, waypoint_category, parent_category: Spatial, parent_category2d: Node2D, collapsed: bool):
+	if waypoint_category.get_name() == "":
+		# If this is called, there is an empty Category in the Waypoint data
 		print("Category found with no name.")
 		return
-	category_item.set_text(0, category.get_display_name())
-	category_item.set_metadata(0, full_category_name)
+
+	var godot_category = category_scene.instance()
+	var godot_category2d = category2d_scene.instance()
+	parent_category.add_subcategory(godot_category)
+	parent_category2d.add_subcategory(godot_category2d)
+
+	var category_item: TreeItem = self.marker_packs.create_item(item)
+	var category_data = CategoryData.new()
+	category_data.waypoint_category = waypoint_category
+	category_data.category = godot_category
+	category_data.category2d = godot_category2d
+
+	category_item.set_metadata(0, category_data)
+	category_item.set_text(0, waypoint_category.get_name())
 	category_item.set_cell_mode(1, TreeItem.CELL_MODE_CHECK)
-	category_item.set_checked(1, Settings.local_category_data.get(full_category_name, {}).get("checked", false))
+	# TODO 214: The format for the category name stored here is a/b/c. 
+	# This could be changed to some UUID.
+	godot_category.name = waypoint_category.get_name()
+	var relative_path: String = self.categories.get_path_to(godot_category)
+	category_item.set_checked(1, Settings.local_category_data.get(relative_path, {}).get("checked", false))
 	category_item.set_tooltip(1, "Show/Hide")
 	category_item.set_editable(1, true)
 	category_item.set_collapsed(collapsed)
 	category_item.set_selectable(1, false)
 
-	for path in category.get_trail():
+	category_data.is_visible = category_item.is_checked(1)
+	godot_category.visible = category_data.is_visible
+	godot_category2d.visible = category_data.is_visible
+
+	for path in waypoint_category.get_trail():
 		var path_points := PoolVector3Array()
 		var trail_data = path.get_trail_data()
 		if trail_data.get_points_x().size() != trail_data.get_points_y().size() or trail_data.get_points_x().size() != trail_data.get_points_z().size():
-			print("Warning: Trail ", full_category_name, " does not have equal number of X, Y, and Z coordinates.")
+			print("Warning: Trail ", waypoint_category.get_name(), " does not have equal number of X, Y, and Z coordinates.")
```

Comment:
This is a good point but I'm trying to avoid keeping the `a.b.c` naming structure. As i mentioned in the TODO comment, the name should be changed to a UUID in the future so we won't be able to use the name long term. Do you have any suggestions?

Reply:
As discussed, we are keeping this as `.get_name` for now.